Obstacle detector

amiller27 opened the pull request:
Tested and works in the sim, only very minor changes since testing on the drone so it should work there too.

Performance is still trash on the drone but fine on my laptop.  This is likely because numpy seems to use SSE2 but not NEON :disappointed:

amiller27 committed:
testing

amiller27 committed:
Added everything

amiller27 committed:
Fixed bug with quaternion component order

amiller27 committed:
Cleaned up obstacle detector

aftersomemath committed:
Fix message type used by obstacle detector, silence warning in orientation filter

amiller27 committed:
Misc cleanup to obstacle detector

asaba96 commented:
I’m going to have more comments later, but 2 big things I see off the bat: 1) is the performance on the drone still acceptable? And 2) can you please add in some comments? It’s very lacking in comments and while most of the math can be gleaned from reading it it’d still make reviewing it and reading it months later much easier if there were some.

Riteshlogin commented on param/obstacle_detector.yaml:
```diff
@@ -0,0 +1,15 @@
+min_x: -5.0
```
Have one comment saying what units these are in and another saying what cluster_threshold is. Everything else in this file is pretty self-evident.

Riteshlogin replied:
Also, because this shows up under Andrew's comment - get_rotation_matrices is especially indecipherable

amiller27 replied:
Fixed

Riteshlogin replied:
Haha, I had to google mks units and I'm surprised I hadn't heard of that before. Thank you!

aftersomemath commented on src/iarc7_sensors/obstacle_detector.py:
```diff
@@ -0,0 +1,358 @@
+import tf2_ros
+import uuid
+
+def slerp(q0, q1, t):
```
Lol I have to jump on the comment bagwagon here too. The wikipedia explanation of slerp has the right amount of comments.

amiller27 replied:
Fixed

aftersomemath commented on src/iarc7_sensors/obstacle_detector.py:
```diff
@@ -0,0 +1,358 @@
+    else:
+        rospy.logerr('Rejecting scan with old timestamp: %s'%msg.header.stamp)
+
+def get_circle(points, radius, drone_pos, settings):
```
Rename to get_cluster_bounding_circle or similar?

amiller27 replied:
Fixed

aftersomemath commented on src/iarc7_sensors/obstacle_detector.py:
```diff
@@ -0,0 +1,358 @@
+    assert points.size
+
+    mean_point = np.mean(points, axis=0)
+    guess_pos_offset = mean_point - drone_pos
```
Can we merge this with the expression below or rename variables so that variable reuse isn't occuring?

amiller27 replied:
Fixed

aftersomemath commented on src/iarc7_sensors/obstacle_detector.py:
```diff
@@ -0,0 +1,358 @@
+    mean_point = np.mean(points, axis=0)
+    guess_pos_offset = mean_point - drone_pos
+    guess_pos_offset *= (
+            (linalg.norm(guess_pos_offset) + settings['obst_radius'])
```
Does this do what its supposed to do? Because it simplifies to guess_pos_offset * (1 + obst_radius/||guess_pose_offset||). The current version calculates the norm twice instead of once.

amiller27 replied:
Yep, it's taking a vector to the boundary of the obstacle and lengthening it into a vector to the center of the obstacle.  I think it's right

aftersomemath reviewed (commented):
Please more comments.

amiller27 committed:
Added some comments and made things clearer

amiller27 commented:
Added some comments and cleaned things up, should be good now.  Performance on the drone is pretty bad, it does run but it hogs a lot of cpu that could be used for other things.  Either way, we won't be running this at competition, and if we were we'd run it on the NUC where it would be blazing fast because SSE2

asaba96 commented on param/obstacle_detector.yaml:
```diff
@@ -0,0 +1,32 @@
+
+# Parameters describing the obstacles
+obst_radius: 0.05
+obst_height: 3.0
```
If this is in meters, why so high? Is this for safety reasons now and eventually, when we have depth cameras, we will actual find the heights?

amiller27 replied:
Yeah this version of the obstacle detector doesn't have any concept of how tall the obstacles are so I just made them as tall as possible.  Future versions will account for height

asaba96 commented on src/iarc7_sensors/obstacle_detector.py:
```diff
@@ -0,0 +1,468 @@
+
+    timeout = rospy.Duration(rospy.get_param('~timeout'))
+
+    rate = rospy.Rate(20)
```
Can we have the rate be a loaded in param?

amiller27 replied:
Fixed

asaba96 reviewed (commented):
Looks very good, appreciate the comments very much!

amiller27 committed:
Made obstacle detector rate a param

asaba96 reviewed (approved):
Everything looks great!
